design/loadbalancer-hash-policy-design.md

```diff
@@ -0,0 +1,185 @@
+Depending on the `RequestAttribute`, some fields of this generic map may be required, an if missing, the hash policy excluded.
+For example, to implement header hashing, the `headerName` field would be required.
+The `Terminal` field denotes if the attribute specified (e.g. header name) is found, Envoy should stop processing any subsequent hash policies in the list it is provided.
+This is a performance optimization and can provide a speedup in hash calculation if set and the attribute to hash on is found.
+
+```
+type RequestHashPolicy struct {
+	RequestAttribute string            `json:"requestAttribute,omitempty"`
+	HashOptions      map[string]string `json:"hashOptions,omitempty"`
+	Terminal         bool              `json:"terminal,omitempty"`
+}
+```
+
+### HTTPProxy YAML Example
+
+An example of how this feature would be used to hash on a specific header value follows below:
+
+```
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+   name: example
+spec:
+  virtualhost:
+    fqdn: example.projectcontour.io
+  routes:
+  - services:
+    - name: example-app
+      port: 80
+    loadBalancerPolicy:
+      strategy: RequestHash
+      requestHashPolicies:
+      - requestAttribute: Header
+        hashOptions:
+          headerName: X-Some-Header
+        terminal: true
+      - requestAttribute: Header
+        hashOptions:
+          headerName: User-Agent
+```
+
+The value of the `X-Some-Header` header will be hashed by Envoy if present and used to make a load balancing decision.
+If the header is present, Envoy will not move on to attempting to hash the `User-Agent` header value.
+Consistent values in this header would lead to Envoy routing to the same backend service instance.
+
+### Handling Invalid Strategy Choice
+As per the documentation of the [`LoadBalancerPolicy` field on the `ExtensionService` object](https://projectcontour.io/docs/v1.11.0/config/api/#projectcontour.io/v1alpha1.ExtensionService), the `Cookie` load balancing strategy is invalid for use with the `ExtensionService`.
+Currently, that restriction does not seem to be enforced (or on the `TCPRoute` object either, for which the `Cookie` strategy is not compatible).
+This design proposes that we start to enforce these strategy restrictions and also restrict the new `RequestHash` strategy to only valid on `Route` objects.
+If specified on `ExtensionService` or `TCPRoute` objects, it will be overridden to the default `RoundRobin` strategy and a warning will be logged.
+Overriding rather than throwing an error is a deliberate attempt to continue to allow the configured `HTTPProxy` or configured extension to function, albeit with basic load balancing enabled.
+
+## Alternatives Considered
+
+### Per-Hash Policy Structs
```

Comment:
I would definitely consider choosing this alternative over using a `map[string]string`. I'm doing something similar for global rate limiting, where there are a number of different ways of generating a rate limit descriptor, and each one has its own set of configuration, and I am planning to have a strongly-typed config struct for each. So choosing this alternative would be consistent with the rate limiting approach. Also, I worry about if there's ever a case where a hash policy requires nested config that doesn't fit nicely into a `map[string]string`. And, like you said, the possible set of these is bounded, it shouldn't be too much of a maintenance burden to add new structs as needed.

Reply:
The nice property of the `map[string]string` is that it allows us to implement the more general "list of hash policies" setup (that is, #3044 ), by adding more options for the `RequestAttribute` field, which all require their own set of options. As you say, there is a bounded set, but I don't think there's also a lot of configuration options for each hash policy.

It does have the disadvantage that we can't do input validation of the fields using the CRD validation bits though, which is a nice user experience.

I think there's enough tradeoffs either way that I don't really have a strong opinion which we pick.

Reply:
I chose the `map[string]string` mostly because of recent chats we've had and misgivings around proliferating too many types, structs, etc., but would probably choose the struct option without that, will switch it around to that

Reply:
Let's see what @stevesloka thinks. I think either one's workable, though I do lean towards the structs.

Reply:
I typically like structs since it's more clear sometimes as to what's going on and can validate better.  If the data lends itself to a struct then 👍  from me.

Reply:
See https://github.com/projectcontour/contour/pull/3268/commits/4852bc08ee0b567204dba03a7a24a6f7ccd9861e